### PR TITLE
feat(idempotency): Add raise_on_no_idempotency_key flag

### PR DESCRIPTION
**Pull request by michaelbrewer**
## Description of changes:

Add an option to raise an error when idempotency key value is None

GIVEN a persistence_store with event_key_jmespath = `body` and option `raise_on_no_idempotency_key`
WHEN getting the hashed idempotency key with an event without a `body` key value
THEN raise IdempotencyKeyError

**Checklist**

 

* [X] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [X] Update tests
* [X] Update docs
* [X] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

## Breaking change checklist

Raising any kind of error when not passing in the data for the idemptency data is not what we currently do

### CODE REVIEW

**Comment by michaelbrewer** [comment]
@cakepietoast  @heitorlessa  - you might want to debate this.

For example when `event_key_jmespath = "[body, queryStringParameters]"`
and the event is `{}` then the `data` is `[None, None]`

**Comment by michaelbrewer** [comment]
> @cakepietoast @heitorlessa - you might want to debate this.
> 
> For example when `event_key_jmespath = "[body, queryStringParameters]"`
> and the event is `{}` then the `data` is `[None, None]`

Also if we do raise an error, should it be a new type for Idempotency key errors?

**Comment by michaelbrewer** [comment]
@heitorlessa @cakepietoast  - i added a `raise_on_no_idempotency_key` option and a new new exception `IdempotencyKeyError`

**Comment by heitorlessa** [comment]
I’ll look into this tomorrow morning and update here as soon as I can.

**Comment by heitorlessa** [comment]
Checking this now

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py [hunk 1]
```diff
@@ -162,8 +168,18 @@ def _get_hashed_idempotency_key(self, lambda_event: Dict[str, Any]) -> str:
...
+        if self.is_missing_idempotency_key(data):
+            warnings.warn(f"No value found for idempotency_key. jmespath: {self.event_key_jmespath}")
+            if self.raise_on_no_idempotency_key:
+                raise IdempotencyKeyError("No data found to create a hashed idempotency_key")
+
         return self._generate_hash(data)
 
+    @staticmethod
+    def is_missing_idempotency_key(data) -> bool:
+        return data is None or (type(data) is list and all(x is None for x in data))
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/persistence/base.py:181]:
```suggestion
        return data is None or not data or all(x is None for x in data)
```

This accounts for empty data structures, including non-lists iterables like tuples, dictionaries

File: tests/functional/idempotency/test_idempotency.py [hunk 2]
```diff
@@ -638,3 +641,44 @@ def test_delete_from_cache_when_empty(persistence_store):
...
+def test_is_missing_idempotency_key():
+    # GIVEN None THEN is_missing_idempotency_key is True
+    assert BasePersistenceLayer.is_missing_idempotency_key(None)
+    # GIVEN a list of Nones THEN is_missing_idempotency_key is True
+    assert BasePersistenceLayer.is_missing_idempotency_key([None, None])
+    # GIVEN a list of all not None THEN is_missing_idempotency_key is false
+    assert BasePersistenceLayer.is_missing_idempotency_key([None, "Value"]) is False
+    # GIVEN a str THEN is_missing_idempotency_key is false
+    assert BasePersistenceLayer.is_missing_idempotency_key("Value") is False
+
```
Comment (heitorlessa) [inline tests/functional/idempotency/test_idempotency.py:655]:
```suggestion
    assert BasePersistenceLayer.is_missing_idempotency_key("Value") is False

    assert BasePersistenceLayer.is_missing_idempotency_key(()) is False
    # GIVEN an empty dictionary THEN is_missing_idempotency_key is false
    assert BasePersistenceLayer.is_missing_idempotency_key({}) is False
```

Add tests for non-lists but iterable too

File: docs/utilities/idempotency.md [hunk 3]
```diff
@@ -259,6 +259,20 @@ payload validation, we would have returned the same result as we did for the ini
 returning an amount in the response, this could be quite confusing for the client. By using payload validation on the
 amount field, we prevent this potentially confusing behaviour and instead raise an Exception.
 
+### Making idempotency key required
+
+By default, events without any idempotency key don't raise any exception and just trigger a warning.
+If you want to ensure that at an idempotency is found, you can pass in `raise_on_no_idempotency_key` as True and an
```
Comment (heitorlessa) [inline docs/utilities/idempotency.md:265]: Not sure what you mean by "If you want to ensure that at an idempotency is found" - could you simplify this to non-native English speakers, please?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
LGTM after logic to catch iterables like empty tuple and dicts as well as doc

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
Merging it - Thanks a lot @michaelbrewer 🎉

**Pull request merged**